clang/lib/Sema/SemaBoundsSafety.cpp

```diff
@@ -186,4 +218,188 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
   return false;
 }
 
+static void EmitIncompleteCountedByPointeeNotes(Sema &S,
+                                                const CountAttributedType *CATy,
+                                                NamedDecl *IncompleteTyDecl,
+                                                bool NoteAttrLocation = true) {
+  assert(IncompleteTyDecl == nullptr || isa<TypeDecl>(IncompleteTyDecl));
+
+  if (NoteAttrLocation) {
+    // Note where the attribute is declared
+    // This is an approximation that's not quite right. This points to the
+    // the expression inside the attribute rather than the attribute itself.
+    //
+    // TODO: Implement logic to find the relevant TypeLoc for the attribute and
+    // get the SourceRange from that (#113582).
+    SourceRange AttrSrcRange = CATy->getCountExpr()->getSourceRange();
+    S.Diag(AttrSrcRange.getBegin(), diag::note_named_attribute)
+        << CATy->getAttributeName(/*WithMacroPrefix=*/true) << AttrSrcRange;
+  }
+
+  if (!IncompleteTyDecl)
+    return;
+
+  // If there's an associated forward declaration display it to emphasize
+  // why the type is incomplete (all we have is a forward declaration).
+
+  // Note the `IncompleteTyDecl` type is the underlying type which might not
+  // be the same as `CATy->getPointeeType()` which could be a typedef.
+  //
+  // The diagnostic printed will be at the location of the underlying type but
+  // the diagnostic text will print the type of `CATy->getPointeeType()` which
+  // could be a typedef name rather than the underlying type. This is ok
+  // though because the diagnostic will print the underlying type name too.
+  // E.g:
+  //
+  // `forward declaration of 'Incomplete_Struct_t'
+  //  (aka 'struct IncompleteStructTy')`
+  S.Diag(IncompleteTyDecl->getBeginLoc(), diag::note_forward_declaration)
+      << CATy->getPointeeType();
+}
+
+static std::tuple<const CountAttributedType *, QualType>
+HasCountedByAttrOnIncompletePointee(QualType Ty, NamedDecl **ND) {
+  auto *CATy = Ty->getAs<CountAttributedType>();
+  if (!CATy)
+    return {};
+
+  // Incomplete pointee type is only a problem for
+  // counted_by/counted_by_or_null
+  if (CATy->isCountInBytes())
+    return {};
+
+  auto PointeeTy = CATy->getPointeeType();
+  if (PointeeTy.isNull())
+    return {}; // Reachable?
```

Comment:
```suggestion
  auto PointeeTy = CATy->getPointeeType();
  if (PointeeTy.isNull())
    return {}; // Reachable?
```
Does a `CountAttributedType` *always* desugar to a `PointerType`? If so, then I don’t think this is reachable because `getPointeeType()` ignores type sugar. In that case, we might want to add `assert(!PointeeTy.isNull())` here?

Reply:
Not always. For a flexible array member like this.

```
struct FAM {
  int size;
  char buffer[] __counted_by(size);
};
```

The type of the `buffer` member is

```
CountAttributedType 0x121079600 'char[] __counted_by(count)' sugar
`-IncompleteArrayType 0x121079500 'char[]' 
  `-BuiltinType 0x121018cb0 'char'
```

so that desugars to an `IncompleteArrayType` rather than a `PointerType`.

I'm not sure if this path is reachable in the contexts that `HasCountedByAttrOnIncompletePointee` is currently called on. It is an error to try to make a flexible-array-member with an incomplete type.

```
struct IncompleteTy;
struct FAM {
  int size;
  struct IncompleteTy buffer[] __counted_by(size); // error
};
```

`HasCountedByAttrOnIncompletePointee` wasn't reachable when I tried simple examples.

That being said we may want to call `HasCountedByAttrOnIncompletePointee` in different contexts in the future so I think returning rather than asserting is the right thing to do here.

Reply:
I left a comment in the code about the case where the early return is reachable.